### ISSUE TITLE
Remove package.json configuration in favor of an rc file

### ISSUE THREAD

**Issue by eliperelman** [mozilla-neutrino/neutrino-dev#199]
Using package.json for overrides can be quite limiting. I'd like to consider having a `neutrinorc.js` (name negotiable of course) file which is a merging of "simple" and "advanced" configuration overrides. We could keep pretty much everything the same, with the benefit that simpler configuration could utilize JS instead of being limited to JSON.

It may seem odd, but we could vary how this file works based on the export.

```js
module.exports = {
  use: [alpha, beta, gamma],
  options: {...},
  config: {...}
};
```

```js
module.exports = neutrino => {
  // ...
};
```

Basically if the exports are an object, we could treat them very similarly to how we do with the package.json content. If the exports are a function, it can just be a piece of middleware to load. The tricky part is determining precedence. Does the rc file take precedence over `--use` middleware, or vice versa?

Going this route we eliminate the need for users to manually specify in their package.json that they are using an override file, since we can pick it up automatically by name.

Thoughts?

**Comment by timkelty** [comment]
Love this idea. If no one has started work I may pitch in.

@eliperelman Especially given that Neutrino/Webpack chain seems to have be built to _avoid_ big configuration objects, I'd actually be in favor dropping the object/package.json style config all together.

My preference would be to _only_ support this:
```
module.exports = neutrino => {
  neutrino.options = {
      source: './app/assets',
      output: './public/assets',
      entry: './js/index.js',
  };

  neutrino.use(require('neutrino-preset-web');
};
```
in favor of:
```
module.exports = {
  options: {
      source: './app/assets',
      output: './public/assets',
      entry: './js/index.js',
  },
  use: [
    'neutrino-preset-web'
  ],
};
```

Seems just as easy, and more explicit.

It would also clear up where MWs/presets should get thier options from (passed via `.use` vs `neutrino.options.myArbitraryMiddlewareProp.myOption`. If they were always called with `.use`, then they can easily accept options.

Thought on `neutrinorc.js`vs `.neutrinorc.js`?

**Comment by constgen** [comment]
My vision is:
```js
module.exports = {
  use: [alpha, beta, 'gamma', function(neutrino){ ... }],
  options: {...},
  config: {...}
};
```
This is similar to what we have in `package.json` now but with JS runtime possibilities. I think `use` may consume module, module name or middleware function (which is the same module). In my opinion this approach keeps the start steps simple as they are now and gives the more consumer friendly advanced customization.

**Comment by timkelty** [comment]
@constgen Sure, However, I would still want the option to export a function instead, for full access to the `neutrino` api.

**Comment by constgen** [comment]
No problem
*.neutrinorc.js*
```js
module.exports = {
  use: ['alpha', 'beta', 'gamma', './custom-preset.js'],
  options: {...},
  config: {...}
}
```
*custom-preset.js*
```js
module.exports =  function(neutrino){ ... }
```
Is it what you want? It is really close to current implementation.

**Comment by eliperelman** [comment]
> It would also clear up where MWs/presets should get thier options from (passed via .use vs neutrino.options.myArbitraryMiddlewareProp.myOption. If they were always called with .use, then they can easily accept options.

This is **very** compelling.

**Comment by eliperelman** [comment]
@timkelty as far as I know this isn't being worked on yet, so feel free to take it on.

**Comment by timkelty** [comment]
@constgen not quite, I'm saying _.neutrinorc.js_ should directly be able to export a function directly. @eliperelman's [original suggestion](https://github.com/mozilla-neutrino/neutrino-dev/issues/199#issue-226158661) would allow for 2 behaviors: object or function.

I'm ok with that approach, but personally I'd prefer just having 1 interface (exporting a function). I get that an object may seem more "user friendly", but I'm [not convinced it's worth the abstraction](https://github.com/mozilla-neutrino/neutrino-dev/issues/199#issuecomment-299617771).

**Comment by constgen** [comment]
@timkelty it should not. Presets and middlewares should.
If you want to be able to export the function any way you can:
*.neutrinorc.js*
```js
exports.use = [function(neutrino){
  ...
}]
```

**Comment by timkelty** [comment]
@constgen I see your point. I suppose that would satisfy my need.
@eliperelman does that work for you?

**Comment by eliperelman** [comment]
Here are the proposals:

1. Export either a function or an object
2. Export only a function
3. Export only an object

My intention was to have the RC file also be acceptable for use as middleware. If we make that flow more convoluted, it would weaken the experience. I don't see any strong arguments in favor of exporting only an object other than keeping a 1:1 mapping from a legacy package.json configuration. This is going in v6, so I see no need to keep that baggage without a strong argument. That's not to say we *can't* accept the export of an object, just that making that our only option seems inconsistent.

I want to evaluate the merits of option 1 and 2.

One of the reasons why the function-as-middleware paradigm works well is it doesn't limit the options provided to middleware. Babel seems to overcome this issue, albeit at the expense of being able to make easy modifications after the fact. Maybe we can allow exporting an object, but adopt some convention for passing options to the middleware automatically:

```js
module.exports = {
  use: ['alpha', 'beta'],
  options: {
    alpha: {},
    beta: {}
  }
};
```

So we need to answer some questions:

- **What value does exporting an object bring?**
- **Does the value exporting an object bring outweigh the complexity in maintenance and implementation required for two export formats?**
- **Should we allow only exporting a function?**
- **If we allow only function exports, is there a significant risk of adoption because of it?**

I'm open to hearing more arguments as well.

**Comment by timkelty** [comment]
Thanks for the recap @eliperelman.

> What value does exporting an object bring?
I don't really see the value, personally. `neutrino.use()...` seems just as easy for the end user, and having 1 interface seems simpler to digest (and maintain).

> If we allow only function exports, is there a significant risk of adoption because of it?
Considering the audience (someone authoring JS w/ webpack), I have a hard time imagining someone being uncomfortable with exporting a function. A proper example would be enough to get someone rolling.

**Comment by constgen** [comment]
I want to remind to everyone that currently (and in the future) the customization of the Webpack Dev Server is not possible in an advanced mode. The only correct way to do this is to set `neutrino.options`. Manual changes of the internal state of the webpack config may break things because a consumer may not know how `port` or `https` params are used. They may be used in multiple spots of the webpack config that definitely will be not considered. He simply may break HMR and etc.
So as for me the declarative way to setup Neutrino in which we can control how and when settings will be applied should be with the highest priority. Imperative way should be avoided as much as possible. I vote for *object only*.

**Comment by timkelty** [comment]
> Server is not possible in an advanced mode. The only correct way to do this is to set neutrino.options.

@constgen I'm not sure I follow.

```
module.exports = neutrino => {
  neutrino.options(...);
};

```

**Comment by constgen** [comment]
Actually now it looks like this 
```js
neutrino.options.server = {
  port: 4000,
  https: false
}
```
What disadvantages I see here:
1. You need to make sure this options are applied before the `neutrino-preset-web`.
2. There is a chance to clear some previous options when actually we want to merge them.

These all are problems of imperative approach.

**Comment by eliperelman** [comment]
Precedence is the hardest thing at the moment. I was hoping a `.neutrinorc.js` file would take the place of an advanced override file needing to be manually specified, but it does suffer from the same issue of options not taking precedence, e.g.:

```js
// neutrino build --use alpha beta

neutrino.options.entry = 'app.js';
// the entry doesn't apply to alpha or beta
```

Exporting an object is one way to overcome this:

```js
const alpha = require('alpha');
const beta = require('beta');

const gamma = neutrino => {
  console.log(neutrino.options.entry); // -> /project/src/app.js
};

module.exports = {
  use: [alpha, beta, gamma],
  options: {
    entry: 'app.js'
  }
};
```

Another option would be to provide some way for middleware to execute code prior to configuration. This would probably mean a significant API change since we'd be effectively eventing this operation:

```js
module.exports = neutrino => {
  neutrino.on('prerun', () => {
    neutrino.options.entry = 'app.js';
  });

  neutrino.on('config', config => {
    config.whatever //...
  });
};
```

I'd be open to other ideas too.

**Comment by timkelty** [comment]
I'd be fine with either of those solutions (allowing object export or providing events).

I like the eventing approach because it solves the precedence problem for middlewares as well, not just `.neutrinorc.js`. One drawback may be that our `.neutrinorc.js` gets a more complicated.

@eliperelman if we went the object route, would you still want to allow a function so `.neutrinorc.js` was usable as a middleware, or allow _object only_ (as @constgen suggests)? It seems more consistent to me that `.neutrinorc.js` could be a middleware, but it would also be nice to only support one interface.

**Comment by constgen** [comment]
Events seem overcomplicated for such things like setting of document title.

This is more friendly for those who just want the work to be done:
```
exports.options.html.title = 'My awesome app'
```
I think that we have to provide the object to use those configuration interfaces that presets and middlewares expose. Authors should explicitly define what can be customized and what not. This is done now but have to be used more often in the future. This means that we can expect more options in the next version of Neutrino. Options need to become a public API for presets.

For real Webpack architectors we have the good old middleware API where you are free to break everything you want.

So I still stay on my side. Declaration is our everything. But need to think about the possibility to configure for different NODE_ENV like somebody mentioned once. May be somehow to expose it to `process` namespace just in the `.neutrinorc.js`?

**Comment by timkelty** [comment]
> Authors should explicitly define what can be customized and what not. This is done now but have to be used more often in the future. This means that we can expect more options in the next version of Neutrino.

I guess I was hoping by dropping `package.json` config, we could encourage middlewares to accept their own options, rather than piggybacking off `neutrino.options`.

For example, since `neutrino-preset-web` uses `neutrino.options.html`.
Why not just have `neutrino-preset-web` accept options directly?

_as function_

```
module.exports = neutrino => {
  neutrino.options.entry = 'app.js';
  neutrino.use(require('neutrino-preset-web'), { html: { title: 'My awesome app' } } );
};
```

_as object_

```
module.exports.options.entry = 'app.js';
module.exports.use = [function(neutrino) {
    neutrino.use(require('neutrino-preset-web'), { html: { title: 'My awesome app' } } );
}];
```

**Comment by constgen** [comment]
I would like it if there will be some options that can be reused in different middlewares at the same time. For example `browserList` for Babel and PostCSS. 

I'm trying to protect interest of people who want to grab package.json, drop it into a project and everything is working magically. What you suggest requires you to program much. This brings us back to the Webpack core issue - you need to spend a weekend to setup a project before you start.

This is a reason why I am here. Ones Neutrino saved my ass.

@timkelty probably I can't convince you. We need more votes in this question.

**Comment by timkelty** [comment]
@constgen I agree my examples aren't drop-in approachable, which is a problem.
I'm far from sold on my suggestions, I'm just exploring our options.

> I would like it if there will be some options that can be reused in different middlewares at the same time. For example browserList for Babel and PostCSS.

Agreed.
It would become important where those options that are shared are documented, however. E.g. would `browserList` become a [core neutrino option](https://neutrino.js.org/customization/simple.html#overriding-neutrino-options)? Seems like it would need to be for middleware authors to know where to reference it.

**Comment by eliperelman** [comment]
`neutrino.options.compile.targets`?

**Comment by constgen** [comment]
Well, yes this is how it is namespaced now. But we could do it better.

**Comment by eliperelman** [comment]
I don't want to get too off-topc here, but it's called `compile.targets` rather than `browserlist` because it needs to make sense for Node.js too.

**Comment by eliperelman** [comment]
You are both making excellent points. You're making this very hard on me 😆.

Maybe we just need a different middleware format. Maybe all middleware can either be a function or an object.

```js
module.exports = ({
  options: { /* options are always merged prior to using more middleware */ },
  // getting very babel-like
  use: [
    'alpha',
    ['beta', { options }],
    neutrino => {
      // whatever
    }
  ]
});
```

```js
module.exports = neutrino => {
  // whatever
};
```

**Comment by timkelty** [comment]
This kind of illustrates the point a bit - suppose I'm writing a postcss middleware and want to pass browserlist options. Where/how do I know about `neutrino.options.compile.targets`, and that it is intended to be a shared option? (as [it isn't documented as such](https://neutrino.js.org/customization/simple.html#overriding-neutrino-options))

**Comment by eliperelman** [comment]
https://neutrino.js.org/presets/neutrino-preset-web/#compile-targets
https://neutrino.js.org/presets/neutrino-preset-node/#compile-targets

**Comment by timkelty** [comment]
Ah ok - so its documented within each of the middlewares, not as a "core" neutrino option. Gotcha.

**Comment by timkelty** [comment]
> Maybe we just need a different middleware format. Maybe all middleware can either be a function or an object.

I'm liking that idea. Gives us parity between middlewares and `.neutrinorc.js`, which seems like a good choice.

**Comment by eliperelman** [comment]
The semi-declarative format is nice, while still allowing for advanced work. With this approach, do we even need a `config` property on the object?

**Comment by timkelty** [comment]
> I'm also thinking neutrino.options can be reserved for API-only options, while everything else should just get passed options. If you want something shared, it could be in a variable?

Yes, please!

So would `neutrino.options.compile.targets` be considered an API option, or be converted to a passed option?

**Comment by eliperelman** [comment]
If we go this route, it would be a passed option. Anything middleware-specific should be middleware arguments.

**Comment by timkelty** [comment]
The "semi-declarative" format gets my vote.

> Maybe all middleware can either be a function or an object.

Just spitballing...if we go that route, should we _only_ support the object format? (going back to @constgen's original suggestion)

**Comment by constgen** [comment]
Little more steps and we will reinvent Grunt

**Comment by timkelty** [comment]
Just following up to see where we're all at with this.

Of the suggestions, I'm partial to @eliperelman's "semi-declarative" format : https://github.com/mozilla-neutrino/neutrino-dev/issues/199#issuecomment-300657868

- Allows a simple object for basic needs, but still allows middlewares to accept options if needed
- Allocates `neutrino.options` for API-only options, not ad-hoc options for middlewares. Those should be passed as options to the middleware.
- If defined as an acceptable middleware format, `.neutrinorc.js` wouldn't need any special parsing, it would be an automatically `use`d middleware if present.
- By using the object format (and perhaps recommending it), we get around the precedence issue, and can always apply `options` early. One could still export a function at their own peril, but should be aware of issues.

> Little more steps and we will reinvent Grunt
@constgen I take it you think this approach is still too complicated?

**Comment by jaridmargolin** [comment]
Only issue I see with passing options in, which I originally thought would be the best approach, is it requires knowledge of what plugins need what options. If I want to change the devServer port, or browserList, I have to know what middleware depend on the shared option and ensure I pass it to each.

Whether or not a "semi-declarative" approach is supported, I don't see any reason not to support exporting a fn and treating it as a preset. If you want to play in "advanced mode" there should be nothing stopping you from doing so. And yes, one could just return an object that points to a custom preset, but why make them jump through hoops?

**Comment by timkelty** [comment]
@jaridmargolin Agreed on allowing a fn export.

> it requires knowledge of what plugins need what options

Very true. I'm ok with the tradeoff, as you'd know by setting some option exactly what middleware you're targeting.

It's also possible that it might make sense to "elevate" certain options to be API options (eg browserList/options.compile.targets) for common options that are likely to be reused. They could still be explicitly set but could pull their defaults from `neutrino.options`.

**Comment by constgen** [comment]
I really don't see the way to reconfigure the Dev Server in such approach. Where should I pass custom options? To preset? Should preset re-expose all its middleware configuration interfaces? Or should I call middlewares directly some how? There is now answer. 
In perfect world there should be no options but only different presets for different cases in your life.

**Comment by eliperelman** [comment]
If you are using simple middleware, then it should be obvious which options you will be passing, e.g. if you consume the devServer middleware directly:

```js
module.exports = {
  use: [
    ['neutrino-middleware-devserver', {
      port: 3000
    }]
  ]
};
```

Higher-level middleware needs to make a conscious decision about what options it may want to still expose. For example, I suppose that the Web preset will still want to expose devServer options, which it can expose at `devServer`:

```js
module.exports = {
  use: [
    ['neutrino-preset-web', {
      devServer: {
        port: 3000
      }
    }]
  ]
};
```

This comes down to documentation of the middleware you are using. This isn't fundamentally different than what we are already doing right now with the eslint middleware and the airbnb middleware. The airbnb re-exposes the options for the eslint middleware.

@timkelty I think we have enough information to move forward with a prototype if you would like to get a PR going!

**Comment by constgen** [comment]
Just want to ask for the last thing:
```js
module.exports = {
  use: [
    ['neutrino-preset-web', {
      devServer: {
        port: 3000
      }
    }]
  ]
};
```
to be
```js
module.exports = {
  use: [
    {
      require: 'neutrino-preset-web',
      options: {
         devServer: {
           port: 3000
         }
      } 
    }
  ]
};
```
Naming is discussible

**Comment by timkelty** [comment]
> @timkelty I think we have enough information to move forward with a prototype if you would like to get a PR going!

I can't guarantee I'll have time, but if I start I will post an update here.

**Comment by timkelty** [comment]
@constgen I think I prefer something like that too.
You might argue that `require` prop could just be `use`, so it would mimic what we're doing at the top level:

```
module.exports = ({
  options: { /* options are always merged prior to using more middleware */ },
  use: [
    'alpha',
    {
      use: 'neutrino-preset-web',
      options: {
        devServer: {
          port: 3000
        }
    },
    neutrino => {
      // whatever
    }
  ]
});
```

So our `use` array could take a `string`, an `object` (with `use`, `options`), or a function.

**Comment by constgen** [comment]
Also not bad idea

**Comment by eliperelman** [comment]
Resolved in #233.
